File: app/src/main/java/org/wikipedia/analytics/CreateAccountFunnel.java
package org.wikipedia.analytics;

import androidx.annotation.NonNull;

import org.json.JSONObject;
import org.wikipedia.WikipediaApp;

public class CreateAccountFunnel extends Funnel {
    private static final String SCHEMA_NAME = "MobileWikiAppCreateAccount";
    private static final int REVISION = 20709917;

    private final String requestSource;

    public CreateAccountFunnel(WikipediaApp app, String requestSource) {
        super(app, SCHEMA_NAME, REVISION);
        this.requestSource = requestSource;
    }

    public void logStart(String loginSessionToken) {
        log(
                "action", "start",
                "loginSessionToken", loginSessionToken
        );
    }

    public void logCaptchaShown() {
        log(
                "action", "captchaShown"
        );
    }

    public void logCaptchaFailure() {
        log(
                "action", "captchaFailure"
        );
    }

    public void logError(String code) {
        log(
                "action", "error",
                "errorText", code
        );
    }

    public void logSuccess() {
        log(
                "action", "success"
        );
    }

    @Override
    protected JSONObject preprocessData(@NonNull JSONObject eventData) {
        preprocessData(eventData, "source", requestSource);
        return super.preprocessData(eventData);
    }
}

File: app/src/main/java/org/wikipedia/analytics/CreateAccountFunnel.kt
package org.wikipedia.analytics

import org.json.JSONObject
import org.wikipedia.WikipediaApp

class CreateAccountFunnel(app: WikipediaApp, private val requestSource: String) : Funnel(app, SCHEMA_NAME, REVISION) {

    fun logStart(loginSessionToken: String?) {
        log("action" to "start", "loginSessionToken" to loginSessionToken)
    }

    fun logError(code: String?) {
        log("action" to "error", "errorText" to code)
    }

    fun logSuccess() {
        log("action" to "success")
    }

    override fun preprocessData(eventData: JSONObject): JSONObject {
        preprocessData(eventData, "source", requestSource)
        return super.preprocessData(eventData)
    }

    companion object {
        private const val SCHEMA_NAME = "MobileWikiAppCreateAccount"
        private const val REVISION = 20709917
    }
}
